k8s/k8s.go

```diff
@@ -0,0 +1,1291 @@
+	// UpdateVolumeAttachmentStatus updates a volume attachment status
+	UpdateVolumeAttachmentStatus(*storagev1beta1.VolumeAttachment) (*storagev1beta1.VolumeAttachment, error)
+}
+
+// SecretOps is an interface to perform k8s Secret operations
+type SecretOps interface {
+	// GetSecret gets the secrets object given its name and namespace
+	GetSecret(name string, namespace string) (*v1.Secret, error)
+	// CreateSecret creates the given secret
+	CreateSecret(*v1.Secret) (*v1.Secret, error)
+	// UpdateSecret updates the given secret
+	UpdateSecret(*v1.Secret) (*v1.Secret, error)
+	// UpdateSecretData updates or creates a new secret with the given data
+	UpdateSecretData(string, string, map[string][]byte) (*v1.Secret, error)
+	// DeleteSecret deletes the given secret
+	DeleteSecret(name, namespace string) error
+}
+
+// ConfigMapOps is an interface to perform k8s ConfigMap operations
+type ConfigMapOps interface {
+	// GetConfigMap gets the config map object for the given name and namespace
+	GetConfigMap(name string, namespace string) (*v1.ConfigMap, error)
+	// CreateConfigMap creates a new config map object if it does not already exist.
+	CreateConfigMap(configMap *v1.ConfigMap) (*v1.ConfigMap, error)
+	// DeleteConfigMap deletes the given config map
+	DeleteConfigMap(name, namespace string) error
+	// UpdateConfigMap updates the given config map object
+	UpdateConfigMap(configMap *v1.ConfigMap) (*v1.ConfigMap, error)
+	// WatchConfigMap sets up a watcher that listens for changes on the config map
+	WatchConfigMap(configMap *v1.ConfigMap, fn core.WatchFunc) error
+}
+
+// ObjectOps is an interface to perform generic Object operations
+type ObjectOps interface {
+	// GetObject returns the latest object given a generic Object
+	GetObject(object runtime.Object) (runtime.Object, error)
+	// UpdateObject updates a generic Object
+	UpdateObject(object runtime.Object) (runtime.Object, error)
+}
+
+// MutatingWebhookConfigurationOps is interface to perform CRUD ops on mutatting webhook controller
+type MutatingWebhookConfigurationOps interface {
+	// GetMutatingWebhookConfiguration returns a given MutatingWebhookConfiguration
+	GetMutatingWebhookConfiguration(name string) (*hook.MutatingWebhookConfiguration, error)
+	// CreateMutatingWebhookConfiguration creates given MutatingWebhookConfiguration
+	CreateMutatingWebhookConfiguration(req *hook.MutatingWebhookConfiguration) (*hook.MutatingWebhookConfiguration, error)
+	// UpdateMutatingWebhookConfiguration updates given MutatingWebhookConfiguration
+	UpdateMutatingWebhookConfiguration(*hook.MutatingWebhookConfiguration) (*hook.MutatingWebhookConfiguration, error)
+	// DeleteMutatingWebhookConfiguration deletes given MutatingWebhookConfiguration
+	DeleteMutatingWebhookConfiguration(name string) error
+}
+
+type privateMethods interface {
+}
+
+var (
+	instance Ops
+	once     sync.Once
+)
+
+type k8sOps struct{}
```

Comment:
This object just comprise of the extracted Ops objects. So when we return a new instance via `NewInstanceFromClients()` or `NewInstanceFromConfigFile()` it will be a brand new object with its own ops clients. We should not set the singleton instances of the individual ops in the `New..` methods.

Reply:
Agree, removing these once.